Guide.docc/CommonProblems.md

```diff
@@ -719,3 +719,65 @@ class WindowStyler {
 All `Sendable` properties can still be safely accessed in this `init` method.
 And while any non-`Sendable` properties cannot,
 they can still be initialized by using default expressions.
+
+### Non-Isolated Deinitialization
+
+By default, any custom initializers you write will have the same isolation as
+their enclosing scope.
+This is _not_ the case for deinitializers, which are always non-isolated.
+
+```swift
+@MainActor
+class WindowStyler {
+    // A MainActor-isolated type
+    private let store = StyleStore()
+
+    init {
+        // infers MainActor
+    }
+
+    deinit {
+        store.stopNotifications()
+    }
+}
+```
+
+This code produces the error:
+
+```
+error: call to main actor-isolated instance method 'stopNotifications()' in a synchronous nonisolated context
+ 5 |     
+ 6 |     deinit {
+ 7 |         store.stopNotifications()
+   |               `- error: call to main actor-isolated instance method 'stopNotifications()' in a synchronous nonisolated context
+ 8 |     }
+ 9 | }
+```
+
+While this might feel surprising, given that this type is `MainActor`-isolated,
+this is not a new constraint.
+The thread that executes a deinitializer has never been guaranteed and
```

Comment:
I'd reword this a bit, we should use this to educate a bit about how deinits work.

```suggestion
Since a deinitializer is synchronous, it is allowed to run inline wherever the last reference to this actor
was released. This is the same as any other reference counted object in Swift works, and deinitializing
the object as soon as possible this way you to have predictable deinitializaiton semantics which are one
of the benefits of automatic reference counting.
```

and let's leave a TODO for us: `deinit isolated` is coming maybe even in 6 this year... so we should then add it here.

Reply:
https://github.com/apple/swift-migration-guide/issues/55